cmd/restic/cmd_forget.go

```diff
@@ -138,14 +138,26 @@ func init() {
 func verifyForgetOptions(opts *ForgetOptions) error {
 	if opts.Last < -1 || opts.Hourly < -1 || opts.Daily < -1 || opts.Weekly < -1 ||
 		opts.Monthly < -1 || opts.Yearly < -1 {
+		// TODO: this seem to have no effect. there is a different error message showing up, e.g. for --keep-last -5
 		return errors.Fatal("negative values other than -1 are not allowed for --keep-*")
 	}
 
+	var allKeepWithinXUndefined = true
 	for _, d := range []restic.Duration{opts.Within, opts.WithinHourly, opts.WithinDaily,
 		opts.WithinMonthly, opts.WithinWeekly, opts.WithinYearly} {
 		if d.Hours < 0 || d.Days < 0 || d.Months < 0 || d.Years < 0 {
 			return errors.Fatal("durations containing negative values are not allowed for --keep-within*")
 		}
+		if d.Hours != 0 || d.Days != 0 || d.Months != 0 || d.Years != 0 {
+			allKeepWithinXUndefined = false
+		}
+	}
+
+	allKeepXUndefined := opts.Last == 0 && opts.Hourly == 0 && opts.Daily == 0 && opts.Weekly == 0 &&
+		opts.Monthly == 0 && opts.Yearly == 0
+
+	if len(opts.KeepTags) > 0 && allKeepXUndefined && allKeepWithinXUndefined {
```

Comment:
Regarding 2.: Why do you think it doesn't cover `--keep-within`? It should be covered by `allKeepWithinXUndefined` which takes `opts.Within` into account.
```
	f.VarP(&forgetOptions.Within, "keep-within", "", "keep snapshots that are newer than `duration` (eg. 1y5m7d2h) relative to the latest snapshot")
```
Should be fine, right?